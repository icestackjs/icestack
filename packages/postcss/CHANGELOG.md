# @icestack/postcss

## 0.6.2

### Patch Changes

- [`854af96`](https://github.com/sonofmagic/icestack/commit/854af960284af72cbf36ebaf6e574c267dea42b1) Thanks [@sonofmagic](https://github.com/sonofmagic)! - add unocss types

- Updated dependencies [[`854af96`](https://github.com/sonofmagic/icestack/commit/854af960284af72cbf36ebaf6e574c267dea42b1)]:
  - @icestack/postcss-utils@0.5.2
  - @icestack/shared@0.6.2
  - @icestack/types@0.6.2

## 0.6.1

### Patch Changes

- [`e89ae02`](https://github.com/sonofmagic/icestack/commit/e89ae02b3d0a826a365e97956923b02429fc74ad) Thanks [@sonofmagic](https://github.com/sonofmagic)! - add cli and fix mode preset issue

- Updated dependencies [[`e89ae02`](https://github.com/sonofmagic/icestack/commit/e89ae02b3d0a826a365e97956923b02429fc74ad)]:
  - @icestack/postcss-utils@0.5.1
  - @icestack/shared@0.6.1
  - @icestack/types@0.6.1

## 0.6.0

### Minor Changes

- [`3311043`](https://github.com/sonofmagic/icestack/commit/33110430aa11e6bd2b58ffcdbd7434df4814c272) Thanks [@sonofmagic](https://github.com/sonofmagic)! - add @icestack/preset-default, change default preset to this package

### Patch Changes

- Updated dependencies [[`3311043`](https://github.com/sonofmagic/icestack/commit/33110430aa11e6bd2b58ffcdbd7434df4814c272)]:
  - @icestack/postcss-utils@0.5.0
  - @icestack/shared@0.6.0
  - @icestack/types@0.6.0

## 0.5.5

### Patch Changes

- [`05c583d`](https://github.com/sonofmagic/icestack/commit/05c583d0683a87069fa9777c847c2784de4336cf) Thanks [@sonofmagic](https://github.com/sonofmagic)! - support raw string in css schema

- Updated dependencies [[`05c583d`](https://github.com/sonofmagic/icestack/commit/05c583d0683a87069fa9777c847c2784de4336cf)]:
  - @icestack/postcss-utils@0.4.17
  - @icestack/shared@0.5.5
  - @icestack/types@0.5.5

## 0.5.4

### Patch Changes

- [`22b607d`](https://github.com/sonofmagic/icestack/commit/22b607d939d2a7c1794f179c6fffed41c7436efb) Thanks [@sonofmagic](https://github.com/sonofmagic)! - add pick options and change mode option

- Updated dependencies [[`22b607d`](https://github.com/sonofmagic/icestack/commit/22b607d939d2a7c1794f179c6fffed41c7436efb)]:
  - @icestack/postcss-utils@0.4.16
  - @icestack/shared@0.5.4
  - @icestack/types@0.5.4

## 0.5.3

### Patch Changes

- [`f48c139`](https://github.com/sonofmagic/icestack/commit/f48c1390111741f21d98c2d8f5117e8ef0d5d080) Thanks [@sonofmagic](https://github.com/sonofmagic)! - add clean options

- Updated dependencies [[`f48c139`](https://github.com/sonofmagic/icestack/commit/f48c1390111741f21d98c2d8f5117e8ef0d5d080)]:
  - @icestack/postcss-utils@0.4.15
  - @icestack/shared@0.5.3
  - @icestack/types@0.5.3

## 0.5.2

### Patch Changes

- [`99ebaf4`](https://github.com/sonofmagic/icestack/commit/99ebaf459d0727f7f57111827362456f2a9f3f27) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: base.mediaDarkTheme option

- Updated dependencies [[`99ebaf4`](https://github.com/sonofmagic/icestack/commit/99ebaf459d0727f7f57111827362456f2a9f3f27)]:
  - @icestack/postcss-utils@0.4.14
  - @icestack/shared@0.5.2
  - @icestack/types@0.5.2

## 0.5.1

### Patch Changes

- [`cd4a6f0`](https://github.com/sonofmagic/icestack/commit/cd4a6f020fea5164011476f61385322a593f994f) Thanks [@sonofmagic](https://github.com/sonofmagic)! - compose postcss options into one

- Updated dependencies [[`cd4a6f0`](https://github.com/sonofmagic/icestack/commit/cd4a6f020fea5164011476f61385322a593f994f)]:
  - @icestack/postcss-utils@0.4.13
  - @icestack/shared@0.5.1
  - @icestack/types@0.5.1

## 0.5.0

### Minor Changes

- [`bcb03ca`](https://github.com/sonofmagic/icestack/commit/bcb03ca5b64672e0467ea047cb8c546a0ad57091) Thanks [@sonofmagic](https://github.com/sonofmagic)! - release 0.5.0

### Patch Changes

- Updated dependencies [[`bcb03ca`](https://github.com/sonofmagic/icestack/commit/bcb03ca5b64672e0467ea047cb8c546a0ad57091)]:
  - @icestack/shared@0.5.0
  - @icestack/types@0.5.0
  - @icestack/postcss-utils@0.4.12

## 0.4.11

### Patch Changes

- [`7a5e4ff`](https://github.com/sonofmagic/icestack/commit/7a5e4ff8f0be55f5377c6c28f11c8badb6f6246a) Thanks [@sonofmagic](https://github.com/sonofmagic)! - ci changes

- Updated dependencies [[`7a5e4ff`](https://github.com/sonofmagic/icestack/commit/7a5e4ff8f0be55f5377c6c28f11c8badb6f6246a)]:
  - @icestack/shared@0.4.11
  - @icestack/types@0.4.11

## 0.4.10

### Patch Changes

- changelog init

- Updated dependencies []:
  - @icestack/shared@0.4.10
  - @icestack/types@0.4.10
