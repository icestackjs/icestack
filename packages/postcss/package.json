{
  "name": "@icestack/postcss",
  "description": "The postcss processor of @icestack/ui",
  "version": "1.1.0",
  "scripts": {
    "build": "unbuild",
    "test": "vitest run",
    "test:dev": "vitest",
    "release": "pnpm publish",
    "sync": "cnpm sync @icestack/postcss"
  },
  "keywords": [
    "postcss",
    "processor",
    "css",
    "component",
    "generate",
    "generator",
    "tailwindcss",
    "unocss",
    "icestack",
    "ui"
  ],
  "sideEffects": false,
  "main": "./dist/index.cjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "dependencies": {
    "@icestack/postcss-utils": "workspace:^",
    "@icestack/shared": "workspace:^",
    "@icestack/types": "workspace:^",
    "css-selector-tokenizer": "^0.8.0",
    "defu": "^6.1.4",
    "lodash": "^4.17.21",
    "postcss": "^8.4.33",
    "postcss-custom-property-prefixer": "workspace:^",
    "postcss-js": "^4.0.1",
    "postcss-scss": "^4.0.9",
    "postcss-selector-parser": "^6.0.15",
    "postcss-value-parser": "^4.2.0",
    "tailwindcss": "^3.4.1"
  },
  "author": "SonOfMagic <qq1324318532@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/sonofmagic/icestack"
  },
  "homepage": "https://ui.icebreaker.top",
  "repository": {
    "type": "git",
    "url": "https://github.com/sonofmagic/icestack"
  }
}
