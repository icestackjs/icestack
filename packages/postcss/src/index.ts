import type { PluginCreator, AcceptedPlugin } from 'postcss'

const creator: PluginCreator<{ a: 1 }> = () => {
  const plugins: AcceptedPlugin[] = [
    // async (root, res) => {
    //   console.log(root)
    // }
    {
      postcssPlugin: 'xxx',
      AtRule(atRule, helper) {
        if (atRule.name === 'icestack') {
          // atRule.params
          // console.log(atRule)
          atRule.params.split('.')
          // path
        }
      }
    }
  ]
  return {
    postcssPlugin: 'postcss-icestack-plugin',
    plugins
  }
}

creator.postcss = true

export default creator
