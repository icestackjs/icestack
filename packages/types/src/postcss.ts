export interface PrefixerOptions {
  prefix?: string
  ignore?: ReadonlyArray<RegExp | string>
}

export type VarPrefixerOptions = {
  varPrefix?: string // | PrefixFunction
  ignoreProp?: (RegExp | string)[]
  ignoreValueCustomProperty?: (RegExp | string)[]
}
