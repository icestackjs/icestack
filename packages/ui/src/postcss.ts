export * from '@icestack/postcss/ast'
export * from '@icestack/postcss/js'
export * from '@icestack/postcss/plugins'
export * from '@icestack/postcss/scss'
export { compressCssSelector, mapCss2JsArray, preprocessCssInJs, recursiveNodes, transformCss2Js } from '@icestack/postcss/utils'
export { AcceptedPlugin, CssInJs, PrefixerOptions, VarPrefixerOptions } from '@icestack/postcss/types'
