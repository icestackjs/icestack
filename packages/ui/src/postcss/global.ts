import type { PluginCreator, AcceptedPlugin, AtRule } from 'postcss'
import parser from 'postcss-selector-parser'
import type { CodegenOptions } from '@/types'
// media(hover:hover)

function isAtMediaHover(atRule: AtRule) {
  return /media\(\s*hover\s*:\s*hover\s*\)/.test(atRule.name) || (atRule.name === 'media' && /\(\s*hover\s*:\s*hover\s*\)/.test(atRule.params))
}
const creator: PluginCreator<CodegenOptions> = (options) => {
  const universal = options?.global?.selector?.universal
  const removeAtMediaHover = !options?.global?.atMedia?.hover
  const universalFn = typeof universal === 'string' ? () => universal : universal
  const ruleTransformer = parser((selectors) => {
    selectors.walk((selector) => {
      // universal
      if (selector.type === 'universal' && selector.value === '*') {
        const s = universalFn?.()
        if (s) {
          selector.value = s
        }
      }
    })
  })
  return {
    postcssPlugin: 'postcss-global-option-handler-plugin',
    plugins: [
      {
        postcssPlugin: 'deep-dark-fantastic',
        AtRule(atRule) {
          // hover

          if (removeAtMediaHover && isAtMediaHover(atRule)) {
            atRule.before(atRule.nodes)
            atRule.remove()
          }
        },
        Rule(rule) {
          ruleTransformer.transformSync(rule, { lossless: false, updateSelector: true })
        }
      }
    ] as AcceptedPlugin[]
  }
}

creator.postcss = true

export default creator
