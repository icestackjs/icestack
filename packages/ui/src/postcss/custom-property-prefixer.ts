import postcss from 'postcss'
import creator, { type UserDefinedOptions } from 'postcss-custom-property-prefixer'
import { defaultVarPrefix } from '@/constants'

export function getPlugin(varPrefix: UserDefinedOptions['prefix'] = defaultVarPrefix) {
  return creator({
    prefix: typeof varPrefix === 'string' ? varPrefix.slice(2) : varPrefix,
    ignoreProp: (decl) => {
      return decl.prop.startsWith('--tw-')
    },
    ignoreValueCustomProperty(customProperty) {
      return customProperty.startsWith('--tw-')
    }
  })
}

export async function addVarPrefix(rawCss: string, varPrefix: string = defaultVarPrefix) {
  const { css } = await postcss([getPlugin(varPrefix)])
    // @ts-ignore
    .process(rawCss, {
      from: undefined
    })
    .async()
  return css
}
