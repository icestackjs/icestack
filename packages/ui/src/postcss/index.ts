import postcssJs from 'postcss-js'
import postcss, { Root, AcceptedPlugin } from 'postcss'
import prefixer from './prefixer'
import type { CodegenOptions } from '@/types'
import type { Options as PrefixerOptions } from '@/postcss/prefixer'
export type { Options as PrefixerOptions } from '@/postcss/prefixer'
export { getPlugin as getCssVarsPrefixerPlugin } from './custom-property-prefixer'
export * from './tailwindcss'

export function resolvePrefixOption(options: string | PrefixerOptions): PrefixerOptions {
  return typeof options === 'string'
    ? {
        prefix: options
      }
    : options
}

export function getPrefixerPlugin(prefix: CodegenOptions['prefix']) {
  if (typeof prefix === 'string') {
    return prefixer({
      prefix
    })
  } else if (typeof prefix === 'object') {
    return prefixer(prefix)
  }
}

export function objectify(root: Root) {
  return postcssJs.objectify(root as Root)
}

export function process(plugins: AcceptedPlugin[], css: string) {
  // @ts-ignore
  return postcss(plugins).process(css, {
    from: undefined
  })
}

export { type Root, type AcceptedPlugin } from 'postcss'
