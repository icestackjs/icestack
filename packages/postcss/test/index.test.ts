import postcss from 'postcss'
import plugin from '@/index'
describe('index', () => {
  it('plugin', () => {
    // @ts-ignore
    const res = postcss([plugin()]).process(`
    @icestack base;
    @icestack components;
    @icestack utilities;
    `)
    expect(res.css).toMatchSnapshot()
  })

  it('plugin', () => {
    // @ts-ignore
    const res = postcss([plugin()]).process(`
    @icestack base.xxxx;
    @icestack components.ddsdsdksaljfds;
    @icestack utilities.dsad.dfsf.ds.fd;
    `)
    expect(res.css).toMatchSnapshot()
  })
})
