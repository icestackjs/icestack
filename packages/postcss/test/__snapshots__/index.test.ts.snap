// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`index > plugin 1`] = `
"
    @icestack base;
    @icestack components;
    @icestack utilities;
    "
`;
